### PR TITLE
Remove injection of atomic JIT functions into NRT memsys.

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title. NRT updated to use ``<stdatomic.h>``.

### CODE REVIEW

**Comment by aseyboldt** [comment]
I *think* the `.h` extension for `atomic` should be there on windows...

**Comment by stuartarchibald** [comment]
> > I _think_ the `.h` extension for `atomic` should not be there on windows...
> 
> Thanks, I think this is indeed the case. I'm also unconvinced that it'll work and that `<atomic>` is C++ only on windows so will have to do a load of subsequent changes.

Sadly it didn't work, it's going to need C++.

**Comment by aseyboldt** [comment]
Well, that is unfortunate...
Might not even be included in their C11 support: https://devblogs.microsoft.com/cppblog/c11-and-c17-standard-support-arriving-in-msvc/#whats-not
Stumbled across this, which might work? I don't know if this requires anything from the linker though. https://docs.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedincrement64

File: numba/core/runtime/nrt.cpp [hunk 1]
```diff
@@ -348,35 +298,35 @@ void NRT_MemInfo_call_dtor(NRT_MemInfo *mi) {
...
 
-void NRT_MemInfo_release(NRT_MemInfo *mi) {
+extern "C" void NRT_MemInfo_release(NRT_MemInfo *mi) {
     NRT_Debug(nrt_debug_print("NRT_MemInfo_release %p refct=%zu\n", mi,
-                                                            mi->refct));
+                                                            mi->refct.load()));
     assert (mi->refct > 0 && "RefCt cannot be 0");
     /* RefCt drop to zero */
-    if (TheMSys.atomic_dec(&mi->refct) == 0) {
+    if ((--mi->refct) == 0) {
```
Comment (gmarkall) [inline numba/core/runtime/nrt.cpp:306]: I always find this hard to evaluate the order of operations (pre-decrement with dereference) - is it worth making this
```suggestion
    if ((--(mi->refct)) == 0) {
```

for people with similar struggles, or is it clear to most?

**Review by gmarkall — COMMENTED** [review]
I'm happy with the general idea of this (after the merge - I merged it locally myself to experiment with but I botched it slightly and made some accidental unrelated edits so I'll have to ask you to do the merge again rather than just offering to push mine).

There are a few small style thoughts / suggestions on the diff, but I don't think it's necessary to follow them if you would prefer not to. I feel like there are some other things we could do to simplify / tidy up the API, but given this is a step on the way to moving this into IR, I think it's not worth suggesting the rearrangements I had in mind given they may not help towards our intended end result.

File: numba/core/runtime/nrt.cpp [hunk 2]
```diff
@@ -41,16 +47,13 @@ static void nrt_fatal_error(const char *msg)
 /*
  * Global resources.
  */
+extern "C" {
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/core/runtime/nrt.cpp]: Thanks, fixed in 6a2389e.

File: numba/core/runtime/nrt.cpp [hunk 3]
```diff
@@ -41,16 +47,13 @@ static void nrt_fatal_error(const char *msg)
...
 
 struct MemSys {
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/core/runtime/nrt.cpp]: Thanks, fixed in 6a2389e.

File: numba/core/runtime/nrt.cpp (hunk 1 above)
Comment (stuartarchibald) [reply numba/core/runtime/nrt.cpp]: Agree, fixed in 6a2389e.

File: numba/core/runtime/nrt.cpp [hunk 4]
```diff
@@ -97,98 +100,45 @@ void NRT_MemSys_set_allocator(NRT_malloc_func malloc_func,
...
  * The MemInfo structure.
  */
 
-void NRT_MemInfo_init(NRT_MemInfo *mi,void *data, size_t size,
+extern "C" void NRT_MemInfo_init(NRT_MemInfo *mi,void *data, size_t size,
                       NRT_dtor_function dtor, void *dtor_info,
                       NRT_ExternalAllocator *external_allocator)
 {
-    mi->refct = 1;  /* starts with 1 refct */
+    mi->refct.store(1);  /* starts with 1 refct */
```
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/core/runtime/nrt.cpp:121]: I think the windows compilers objected to this, I'll give it another try once I've stabilised this branch again.

**Comment by stuartarchibald** [comment]
Buildfarm ID: `numba_smoketest_cpu_yaml_114`.

**Review by gmarkall — DISMISSED** [review]
Thanks for the updates - this looks good, so if all is well on the build farm I think this will be good to merge.

**Comment by stuartarchibald** [comment]
Thanks for the review @gmarkall.

Unfortunately this failed:

> Buildfarm ID: `numba_smoketest_cpu_yaml_114`.

The tests in `numba.tests.test_pycc` failed as they couldn't find `nrt.cpp`, I think this is because the declaration for the distribution was looking solely at `*.c` files which meant the renamed `nrt.c->nrt.cpp` was missing, patch cfd0b03 attempts to fix. I'm pretty sure the reason public CI passes is because it's doing a local build and running tests from that, so the file will exist.

**Comment by stuartarchibald** [comment]
Buildfarm ID: `numba_smoketest_cpu_yaml_115` is running cfd0b03.

**Comment by sklam** [comment]
`numba_smoketest_cpu_yaml_115` passed

**Pull request merged**